gcc/rust/expand/rust-macro-builtins-asm.cc

```diff
@@ -637,9 +740,8 @@ parse_asm (location_t invoc_locus, AST::MacroInvocData &invoc,
 			     is_global_asm == AST::AsmKind::Global);
   auto inline_asm_ctx = InlineAsmContext (inline_asm, parser, last_token_id);
 
-  // operands stream, also handles the optional ","
-  tl::expected<InlineAsmContext, std::string> resulting_context
-    = tl::expected<InlineAsmContext, std::string> (inline_asm_ctx);
+  tl::expected<InlineAsmContext, InlineAsmParseError> resulting_context
+    = tl::expected<InlineAsmContext, InlineAsmParseError> (inline_asm_ctx);
```

Comment:
so AFAICT this is necessary purely to do the monadic operations on line 745, with `resulting_context.and_then(parse_format_strings)`

I think it would be better to have the following:

```c++
parse_format_strings(inline_asm_ctx).and_then(parse_asm_arg).and_then(validate);
```
what do you think?

Reply:
nice catch Arthur, I guess i was afraid of doing that because i ran into a lot of type errors when beginning using tl::expected so i wanted to take it slow. I'm fixing it now